dali/operators/decoder/inflate/inflate_gpu.cc

```diff
@@ -0,0 +1,151 @@
+
+namespace dali {
+
+namespace inflate {
+
+template <typename Status>
+void nvcompCall(Status status) {
+  if (status != nvcompSuccess) {
+    throw std::runtime_error(
+        make_string("Inflate GPU op error: nvComp returned non-zero status: `", status, "`."));
+  }
+}
+
+class InflateOpGpuLZ4Impl : public InflateOpImplBase<GPUBackend> {
+ public:
+  explicit InflateOpGpuLZ4Impl(const OpSpec &spec) : InflateOpImplBase<GPUBackend>{spec} {}
+
+  void RunImpl(Workspace &ws) override {
+    const auto &input = ws.template Input<GPUBackend>(0);
+    auto &output = ws.template Output<GPUBackend>(0);
+    auto total_chunks_num = params_.GetTotalChunksNum();
+    output.SetLayout(params_.GetOutputLayout());
+    SetupInChunks(input);
+    SetupOutChunks(output);
+    auto stream = ws.stream();
+
+    kernels::DynamicScratchpad scratchpad({}, stream);
+    size_t *actual_out_sizes = scratchpad.AllocateGPU<size_t>(total_chunks_num);
+
+    size_t *in_sizes;
+    void const **in;
+    size_t *out_sizes;
+    void **out;
+    std::tie(in_sizes, in, out_sizes, out) = scratchpad.ToContiguousGPU(
+        stream, params_.GetInChunkSizes(), input_ptrs_, inflated_sizes_, inflated_ptrs_);
+
+    size_t tempSize;
+    nvcompCall(nvcompBatchedLZ4DecompressGetTempSize(total_chunks_num, params_.GetMaxOutChunkVol(),
+                                                     &tempSize));
+
+    void *temp = scratchpad.AllocateGPU<uint8_t>(tempSize);
+    nvcompCall(nvcompBatchedLZ4DecompressAsync(in, in_sizes, out_sizes, actual_out_sizes,
+                                               total_chunks_num, temp, tempSize, out, nullptr,
+                                               stream));
+
+    // nvCOMP uses ``out_sizes`` to avoid OOB accesses if the actual data size after the compression
+    // is greater than what follows from the shapes reported by the user.
+    // Here, we take care of the opposite case, when the allocations are bigger than really
+    // necessary and thus may contain some uninitialized memory.
+    FillTheTails(output.type(), total_chunks_num, out, actual_out_sizes, out_sizes, stream);
+  }
+
+ protected:
+  template <typename TL>
+  void SetupInChunks(const TL &input) {
+    auto in_view = view<const uint8_t>(input);
+    auto batch_size = in_view.shape.num_samples();
+    const auto &num_chunks_per_sample = params_.GetChunksNumPerSample();
+    auto total_chunks_num = params_.GetTotalChunksNum();
+    const auto &offsets = params_.GetInChunkOffsets();
+    input_ptrs_.clear();
+    input_ptrs_.reserve(total_chunks_num);
+    for (int64_t sample_idx = 0, chunk_flat_idx = 0; sample_idx < batch_size; sample_idx++) {
+      auto num_chunks = num_chunks_per_sample[sample_idx].num_elements();
+      const uint8_t *sample_base_ptr = in_view.data[sample_idx];
+      for (int chunk_idx = 0; chunk_idx < num_chunks; chunk_idx++, chunk_flat_idx++) {
+        input_ptrs_.push_back(sample_base_ptr + offsets[chunk_flat_idx]);
+      }
```

Comment:
So, you've flattened chunk offsets and now you reconstitute the mapping between samples and chunk offsets.
```suggestion
      auto &&offsets = chunk_offsets_[sample_idx];
      for (int chunk_idx = 0; chunk_idx < num_chunks; chunk_idx++) {
        input_ptrs_.push_back(sample_base_ptr + offsets.data[chunk_idx]);
      }
```